cmd/snap/inhibit.go

```diff
@@ -0,0 +1,152 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/snapcore/snapd/cmd/snaplock/runinhibit"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/progress"
+)
+
+func inhibitMessage(snapName string, hint runinhibit.Hint) string {
+	switch hint {
+	case runinhibit.HintInhibitedForRefresh:
+		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
+	default:
+		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+	}
+}
+
+func isGraphicalSession() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("Snap package waiting for update")
+
+	// Run zenity with a progress bar.
+	// TODO: while we are waiting ask snapd for progress updates and send those
+	// to zenity via stdin.
+	zenityDied := make(chan error, 1)
+
+	// TODO: use a dbus API to allow integration with native desktop environment.
+	cmd := exec.Command(
+		"zenity",
+		// [generic options]
+		"--title="+zenityTitle,
+		// [progress options]
+		"--progress",
+		"--text="+inhibitMessage(snapName, hint),
+		"--pulsate",
+	)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Make sure that zenity is eventually terminated.
+	defer cmd.Process.Signal(os.Interrupt)
+	// Wait for zenity to terminate and store the error code.
+	// The way we invoke zenity --progress makes it wait forever.
+	// so it will typically be an external operation.
+	go func() {
+		zenityErr := cmd.Wait()
+		if zenityErr != nil {
+			zenityErr = fmt.Errorf("zenity error: %s\n", zenityErr)
+		}
+		zenityDied <- zenityErr
+	}()
+
+	if err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, zenityDied); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, nil)
+	pb.Finished()
+	return err
+}
+
+func waitWhileInhibited(snapName string) error {
+	hint, err := runinhibit.IsLocked(snapName)
+	if err != nil {
+		return err
+	}
+	if hint == runinhibit.HintNotInhibited {
+		return nil
+	}
+
+	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+	// and then either unlocks it or changes to HintInhibitedForRefresh (see
+	// gateAutoRefreshHookHandler in hooks.go).
+	// waitInhibitUnlock will return also on HintNotInhibited.
+	if err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh, nil); err != nil {
```

Comment:
so the flow here if the lock is held by snapd to run the gate-auto-refresh hook is that we wait for that hook to finish and either set the inhibit lock to unlocked, or to set it to HintInhibitedForRefresh. Is the expectation that the gate-auto-refresh hook will run in a "short" amount of time? because otherwise if that hook takes a while to process, `snap run` will be stuck here displaying nothing to the user, and only then after the hook is done (at which point maybe the lock is even unlocked if the refresh isn't happening) do we progress and display any indication that the process is blocked waiting for something to happen either graphically with zenity or textually. 

Basically, my concern is about the ordering if we shouldn't also display zenity while waiting for this condition as well as the condition for the refresh hook no longer being held

Reply:
This is addressing Samuele's comment https://github.com/snapcore/snapd/pull/7700#discussion_r659741283 about not displaying anything while waiting for the hook. In theory the hook is expected to call 'snapctl refresh --hold|--proceed' (which itself is quick), but of course it may have arbitrary logic and falls into general hook timeout limit.

Reply:
the issue is that the hook can either say don't proceed or proceed, it's a confusing experience from my POV to show a notification just about the fact that we are running the hook to check whether to proceed and then a different one if we are really inhibited, and we can't really say that we are waiting for the snap to be refreshed if it won't be refreshed either. So under the assumption that the hook is quick (that's under the control of the people creating snaps) this experience seems reasonable. A more refined experience would be to put a notification up if the hook is taking too long, but wait a small reasonable amount of time before doing that. This can be done as an enhancement later on top of the baseline here.

Reply:
@pedronis that explanation makes sense, I agree the best case scenario in this case is that we only show a popup notification if the gate hook takes "too long"